### ISSUE TITLE
Datacat can’t generate target from template.

### ISSUE THREAD

**Issue by abraham1901** [richardc/puppet-datacat#20]
Reproduce steps:

Manifest:
```puppet
datacat { 'test':
  template => 'datacat/test.erb',
  path          => '/tmp/test.txt',
}

datacat_fragment { "${::fqdn} in device hostgroup":
  target => [ 'test' ],
  data   => {
    device => [ $::fqdn ],
  },
}
```

Template:
```
cat modules/datacat/template/test.erb                            
<%= @data["device"].sort.join(",") %>
```

After apply manifest target file is empty.

Content var is empty:

```
      content = template.result(vars.get_binding)
      debug "test2 #{content.inspect}"

Debug: Datacat_collector[test](provider=datacat_collector): test2 ""
```

Ruby reproduce code:
```ruby
require 'erb'

module Test_Bind
    class Datacat_Binding
      def initialize(d, template)
        @data = d
        @__file__ = template
      end

      def file
        @__file__
      end

      def get_binding
        binding()
      end
    end
end

templ = 'datacat/test.erb'
#templ = '/etc/puppet/environments/production/modules/datacat/templates/test.erb'
#templ = 'datacat/templates/test.erb'
data = {"device"=>["webapps-backend1.rg.net"]}


vars = Test_Bind::Datacat_Binding.new(data, templ)

#template = ERB.new('<%= @data["device"].sort.join(",") %>', 0, '-')
template = ERB.new(templ)
template.filename = templ.to_s

puts template.result(vars.get_binding)
```

Result script – empty.

Tested datacat 0.5.0 – 0.6.1.
OS: Ubuntu 14.04.
Puppet master: 3.6.2-1puppetlabs1
Ruby: 1.9.3, 1.9.1, 1.8

**Comment by richardc** [comment]
I think you're failing to understand your bug here.  That your template is called 'datacat/test.erb'  suggests you're adding it to the wrong place.  My module is called datacat, your module with your template should be called something else.

Running your reproduction behaves entirely as I'd expect from reading it:

```
$ ruby repro.rb
datacat/test.erb
```
You set a template body to `datacat/test.erb` and expanded it.

If I change your reproduction script in the following way:

```ruby
require 'erb'

module Test_Bind
    class Datacat_Binding
      def initialize(d, template)
        @data = d
        @__file__ = template
      end

      def file
        @__file__
      end

      def get_binding
        binding()
      end
    end
end

filename = 'datacat/test.erb'

data = {"device"=>["webapps-backend1.rg.net"]}


vars = Test_Bind::Datacat_Binding.new(data, filename)

template = ERB.new('Here is a value: <%= @data["device"].sort.join(",") %>', 0, '-')

puts template.result(vars.get_binding)
```

Then it also behaves as expected:

```
$ ruby repro_fixed.rb
Here is a value: webapps-backend1.rg.net
```

**Comment by richardc** [comment]
Also, even if you have messed with the datacat module as a place to put your templates, you've done it badly.  The path should be datacat/templates/test.erb, not datacat/template/test.erb as you have shown:  https://docs.puppetlabs.com/puppet/latest/reference/modules_fundamentals.html#templates

**Comment by abraham1901** [comment]
Yes. With body template – work body template. 

```ruby
filename = 'datacat/test.erb'
vars = Test_Bind::Datacat_Binding.new(data, filename)

body_template='Here is a value: <%= @data["device"].sort.join(",") %>'
template = ERB.new(body_template)
template.filename = filename.to_s
puts template.result(vars.get_binding)
```

If body_template is empty, template don’t used.
```ruby
body_template=''
filename = 'datacat/test.erb'
vars = Test_Bind::Datacat_Binding.new(data, filename)
template = ERB.new(body_template)
template.filename = filename.to_s

puts template.result(vars.get_binding)
```

Template 'datacat/test.erb' newer used?


>Also, even if you have messed with the datacat module as a place to put your templates, you've done >it badly. The path should be datacat/templates/test.erb, not datacat/template/test.erb as you have >shown: https://docs.puppetlabs.com/puppet/latest/reference/modules_fundamentals.html#templates

Of course you're right, it's my typo.

**Comment by richardc** [comment]
> If body_template is empty, template don’t used.

If it's empty, then it is still used, it's just that the empty input template expands an empty result.

**Comment by abraham1901** [comment]
Ok. How use template?

In code datacat template_body already usage modules/datacat/lib/puppet/provider/datacat_collector/datacat_collector.rb:
```ruby
      template = ERB.new(@resource[:template_body] || '', 0, '-')
```
```ruby
template_body = @resource[:template_body] or ''
```

**Comment by richardc** [comment]
The datacat define takes care of specifying a template body for you:  https://github.com/richardc/puppet-datacat/blob/master/manifests/init.pp#L76-L88

**Comment by abraham1901** [comment]
datacat does work without template body?

**Comment by richardc** [comment]
the datacat define should not work if the file is absent, as the `template_body()` function should fail:  https://github.com/richardc/puppet-datacat/blob/master/lib/puppet/parser/functions/template_body.rb#L4

Please check carefully that your usage is pointing at a template with content.

**Comment by abraham1901** [comment]
Sorry, it’s my fail. 
It’s trouble in old version datacat (0.5.0).
```ruby
   $template_real = $template ? {
-    default => $template,
     undef   => 'inline',
+    default => $template,
   }
 
   $template_body_real = $template_body ? {
-    default => $template_body,
     undef   => template_body($template_real),
+    default => $template_body,
   }
```

Thank you for you help richardc!

**Comment by richardc** [comment]
Fixed in 0.6.1 (already released)
